### PR TITLE
Simplify Http Request Building in E2E Tests

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Please include a summary of the changes.

This bug fixes ... \ This new feature can be used to ...

# Context

Users ran into this bug doing ... \ Users needed this feature to ...

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: test/testutils/http.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	scheme   string
+	hostname string
+	port     uint32
+	path     string
+
+	body string
+
+	host    string
+	headers map[string]string
```
Comment (ben-taussig-solo) [inline test/testutils/http.go:29]: I think we should use `map[string][]string` here. We want to be able to send multi-value headers, and if we do this we can simplify lines 138-140 to `request.Header = h.headers`

File: test/e2e/buffer_test.go [hunk 2]
```diff
@@ -79,7 +61,12 @@ var _ = Describe("buffer", func() {
 			})
 
 			It("valid buffer size should succeed", func() {
-				Eventually(testRequest(), 10*time.Second, 1*time.Second).Should(matchers.HaveExactResponseBody("{\"value\":\"test\"}"))
+				requestBuilder := testContext.GetHttpRequestBuilder().
```
Comment (ben-taussig-solo) [inline test/e2e/buffer_test.go:64]: can we create `requestBuilder` in a broader scope so we don't have to instantiate it in each of these tests?

File: test/e2e/test_context.go [hunk 3]
```diff
@@ -248,8 +248,28 @@ func (c *TestContext) PatchDefaultUpstream(mutator func(us *gloov1.Upstream) *gl
...
 
-// Use a different function to create a test upstream call before testContext.BeforeEach()
+// SetUpstreamGenerator Use a different function to create a test upstream call before testContext.BeforeEach()
 // Used for example with helpers.NewTestGrpcUpstream which has the side effect of also starting a grpc service
 func (c *TestContext) SetUpstreamGenerator(generator func(ctx context.Context, addr string) *v1helpers.TestUpstream) {
 	c.testUpstreamGenerator = generator
 }
+
+// GetHttpRequestBuilder returns an HttpRequestBuilder to easily build http requests used in e2e tests
+func (c *TestContext) GetHttpRequestBuilder() *testutils.HttpRequestBuilder {
```
Comment (ben-taussig-solo) [inline test/e2e/test_context.go:258]: just curious, what's the rationale behind defining `GetHttpRequestBuilder` and `GetHttpsRequestBuilder` here instead of in `test/testutils/http.go`?

File: test/e2e/proxy_protocol_test.go [hunk 4]
```diff
@@ -134,15 +124,15 @@ var _ = Describe("Proxy Protocol", func() {
 		Context("with TLS", func() {
 
 			BeforeEach(func() {
-				requestScheme = "https"
+				requestBuilder = testContext.GetHttpsRequestBuilder()
 				rootCACert = gloohelpers.Certificate()
 
 				secureVsToTestUpstream := gloohelpers.NewVirtualServiceBuilder().
```
Comment (nfuden) [inline test/e2e/proxy_protocol_test.go:130]: Since we are doing this to clean up things with regards to the builder perhaps its also time to consider having these actions not be on the builder interface but rather as options that you pass to the New or with an Apply? 
Also having a copy function might be worthwhile for these as we could then pass around an Anded set of options to make reuse more easily.

For example NewVirtualServiceBuilderFromOptions(WithName(e2e.DefaultVirtualServiceName),
					WithNamespace(writeNamespace))

File: test/testutils/http.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	h.port = port
+	return h
+}
+
+func (h *HttpRequestBuilder) WithPath(path string) *HttpRequestBuilder {
+	h.path = path
+	return h
+}
+
+func (h *HttpRequestBuilder) WithPostBody(body string) *HttpRequestBuilder {
```
Comment (ben-taussig-solo) [inline test/testutils/http.go:77]: Can we just make this `WithBody` and not set the request method here? I feel like there might be situations where we want to set a body but not change the request method (I.E., PUT request).

I could potentially be open to changing the method to POST if the current method is GET, but to be honest I think it would probably just be better to error in that case because I'm not sure if it's ideal for our test utils to have potentially misleading/unexpected behavior

File: test/e2e/proxy_protocol_test.go (hunk 4 above)
Comment (sam-heilbron) [reply test/e2e/proxy_protocol_test.go:130]: Certainly open to it. I'm not entirely sure I follow. How does a constructor that accepts options differ from just a new builder that uses a fluent api and allows you to define the specific options you want?

My hope for a copy was that you can just re-use a builder or extend one:
```
builderEast := [initialize]

builderWest := builderEast.WithName("custom-name")
```

File: test/e2e/test_context.go (hunk 3 above)
Comment (sam-heilbron) [reply test/e2e/test_context.go:258]: Yeah it's a fair question. My hope was to have:
1. A generic utility that we can use across all tests
2. Some specifics for our e2e test

I decided to export "GetHttpRequestBuilder" in the TestContext, because it initializes a request with details that are particular to our e2e tests. Specifically, it sets the port to the port we care about and the host to the host we need to make requests. Previously developers had to always set these 2 things but now they don't have to think about it becomes it comes by default.

File: test/e2e/buffer_test.go (hunk 2 above)
Comment (sam-heilbron) [reply test/e2e/buffer_test.go:64]: Yeah it's a fair point and I'm happy to move it out in the scope. The real hope was the instead of re-using variables across tests, we can just have a test construct the resources it needs locally. Since some tests use http and others https, it seems that re-using some object across a suite might just confuse people. However, in this test we are clearly only testing certain functionality so I'm happy to inititalize it in the broader scope. What do you think?

File: test/testutils/http.go (hunk 5 above)
Comment (sam-heilbron) [reply test/testutils/http.go:77]: Yeah my thinking of PostBody was to try to turn a common setter (set a body, set the method to post) into a single api call. I figured though we don't have it, as you say, we might need to add a `withPutBody` in the future if we need to suport it.

Again, the hope is just ease of use for developers. If you think separating this into 2 calls:
WithBody()
WithMethod()

would be easier, I'm more than happy to adjust

File: test/testutils/http.go (hunk 1 above)
Comment (sam-heilbron) [reply test/testutils/http.go]: https://github.com/solo-io/gloo/pull/7981/commits/aa1a91970e4ab0e3c659d8d86831650b28f40ae7

let me know what you think about this approach. I tried to keep the syntax simple (a single string) because most times people will only set a single value. However, I made it so that you can pass a comma separate list in the case that you need to set multiple values. We don't have test cases that use this yet, but might in the future

File: test/e2e/proxy_protocol_test.go (hunk 4 above)
Comment (sam-heilbron) [reply test/e2e/proxy_protocol_test.go:130]: To be clear, I'm not than happy to shift the API, just wanted to undersatnd what you saw as the benefits to a constructor with a set of option functions that we can pass in, versus just the existing pattern

File: test/testutils/http.go (hunk 1 above)
Comment (ben-taussig-solo) [reply test/testutils/http.go]: This looks good! I feel a little strange about this because `,` is a valid character in HTTP headers. That being said, I don't expect to run into the situation where we intentionally send a header value that contains a literal `,` in our tests so I am resolving the thread

File: test/e2e/proxy_protocol_test.go (hunk 4 above)
Comment (sam-heilbron) [reply test/e2e/proxy_protocol_test.go:130]: whoops! initially wrote that as "not that happy" but meant to write "more than happy". updated.

File: test/e2e/buffer_test.go (hunk 2 above)
Comment (ben-taussig-solo) [reply test/e2e/buffer_test.go:64]: I see your point + I think the git diff tricked me a little bit here -- looking at the file as a whole, I think we probably lose some clarity by defining `requestBuilder` in a different scope. Given how many lines are between some of these tests, I think it's more expressive to just construct the requestBuilder in line

TL;DR on second glance this seems good as is

File: test/e2e/test_context.go (hunk 3 above)
Comment (ben-taussig-solo) [reply test/e2e/test_context.go:258]: ah, makes perfect sense! thanks!

File: test/testutils/http.go (hunk 5 above)
Comment (ben-taussig-solo) [reply test/testutils/http.go:78]: We already have `WithMethod`, so I think it makes sense to just make this `WithBody` and then every one of the `With..` methods only affects a single property on the `HttpRequestBuilder`. I think if `WithPostBody` is the only way to set the body property, it can potentially lead to confusion if someone doesn't realize that they're also modifying the method by calling it

Comment (sam-heilbron) [reply test/testutils/http.go]: Yeah that's fair. What do you think about https://github.com/solo-io/gloo/pull/7981/commits/f3726d8e7bcec2fabd1a965bd69f93147f5df537?

I made sure we expose both WithMethod and WithBody, and left WithPostBody around and a comment that outlines it just aggregates the two

Comment (ben-taussig-solo) [reply test/testutils/http.go]: Yeah that makes sense! I like `WithPostBody` as a convenience method!

File: test/testutils/http_request.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		headers:  make(map[string][]string),
+	}
+}
+
+func (h *HttpRequestBuilder) WithContext(ctx context.Context) *HttpRequestBuilder {
+	h.ctx = ctx
+	return h
+}
+
+func (h *HttpRequestBuilder) WithMethod(method string) *HttpRequestBuilder {
```
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply test/testutils/http_request.go:53]: Yeah that's a great point, ill update

[the remark this answers is not in the document]
Comment (sam-heilbron) [reply test/testutils/http_request.go]: https://github.com/solo-io/gloo/pull/7981/commits/eb011ccccf3fa80a7006c6320b2df8a5d1af856d

**Comment by sam-heilbron** [comment]
/kick https://github.com/solo-io/gloo/issues/7932#issuecomment-1477236420

File: test/e2e/proxy_protocol_test.go (hunk 4 above)
Comment (nfuden) [reply test/e2e/proxy_protocol_test.go:125]: I like the signature more as you can build complex options as a unit with functional statements. That being said its not a big deal. The copy thing is a big deal. 
Unless I am missing something the thing I am worried about extending something is eventually we may have `sharedbuilder := NewBuilder().WithName("blah").WithSomething("else")

    Context(test1, blah){
         test1Builde := sharedBuilder.WithName("east")
    }
    Context(test2, bleegh){
      test2Builde := sharedBuilder.WithSomething("else")
      
     }

`
Because the with extensions modify the builder the name "east" leaks to test2 (or with randomize it may only sometimes leak)

Comment (nfuden) [reply test/e2e/proxy_protocol_test.go:125]: In general the option based approach means that we are more explicit about not passing  the structure just the modifications to a default.
so 
`requestBuilder := testContext.GetHttpRequestBuilder().
					WithPath("1").
					WithPostMethod()`
becomes 
`requestBuilder := testContext.GetHttpRequestBuilder(
					opts.WithPath("1"), 
					opts.WithPostMethod())`

or at times 
`
 sharedOpts := opts.And(opts.WithPath("1"), 
					opts.WithPostMethod()))
requestBuilder := testContext.GetHttpRequestBuilder(sharedOpts)

`

**Review by nfuden — DISMISSED** [review]

**Review by nfuden — APPROVED** [review]

**Review by ben-taussig-solo — APPROVED** [review]
